src/Groups/GAPGroups.jl

```diff
@@ -121,26 +121,29 @@ Return the exponent of `G`, i.e. the smallest positive integer `e` such that `g`
 """
 Base.exponent(x::GAPGroup) = GAP.Globals.Exponent(x.X)
 
+function _gap_rand(f, G::GAPGroup)
+    s = f(G.X)
+    return group_element(G, s)
+end
+
 """
     rand(G::Group)
 
 Return a random element of the group `G`.
 """
-function Base.rand(x::GAPGroup)
-   s = GAP.Globals.Random(x.X)
-   return group_element(x, s)
+function Base.rand(rng::Random.AbstractRNG, rs::Random.SamplerTrivial{Gr}) where Gr<:Oscar.GAPGroup
+   G = rs[]
+   return _gap_rand(GAP.Globals.Random, G)
```

Comment:
Isn't this cheating? Formally, a Julia RNG is entered as an argument, but inside the function a different RNG is used on the GAP side.
`GAP.Globals.Random` calls support a GAP RNG as the first argument. (Yes, the GAP documentation is suboptimal in this respect.)
If we want to admit a *Julia* RNG as the first argument of `Base.rand` where the second argument depends on GAP then I think a correct solution would be to admit a *Julia* RNG as the first argument of `GAP.Globals.Random`. Perhaps it is already sufficient to support `GAP.Globals.Random( rng, from, to )`; GAP.jl is the right place to provide this --I will try.

Reply:
It obviously is cheating, but I don't know enough of GAP↔julia interaction to know what and how should be passed

Reply:
Just to make progress, perhaps we can merge this "cheat version", but with a little comment adding along the lines of
> HACK/FIXME/TODO: we ignore the `rng` and `rs` arguments for now. Find a way to let GAP use them.

One way to do that could be by implementing a `JuliaRandomState` type in GAP that wraps these in a way suitable for GAP.

However, creating and discarding such an object constantly could be a serious issue in terms of performance. My guess is that most of the time, the same `rng` (and the same `rs`, I hope?) will be used. So we could have a `WeakValueDict` somewhere which uses the `rng` (resp. `rs`?) as key and maps that to the GAP wrapper object. Alas... that's kind of a circular dependency... so I am not sure it'd actually work...??? Alternatively, a LRU (least recently used) cache could be used; we'd store the wrappers for the N last rng (or `rs`, or pairs `(rng, rs)`), and when new ones are added, the least recently used one is dropped. See e.g. https://github.com/JuliaCollections/LRUCache.jl

Or perhaps you folks have a better idea, that'd be great :-)

Reply:
I think @rfourquet is the (global) expert on randomness in julia, so I'm conjuring him now :tongue:

Reply:
I do not vote against merging the pull request.

Concerning the use of a Julia RNG in GAP, my idea is as @fingolfin sketches: Wrap the RNG into a GAP RNG that gets handed over to inner `Random` calls in GAP until the other arguments of `Random` are a GAP list or two integers, and then the Julia function gets called.
The answer to the performance problem (don't create a new GAP object for each `rand` call) is caching.

Once this feature is available in GAP.jl, also the reverse direction naturally fits into GAP.jl, that is, the use of GAP random sources in Julia. @rfourquet had proposed code for that in oscar-system/GAP.jl/pull/544.
(After an experiment, I think that putting this code into Oscar.jl instead of GAP.jl is not a good idea; in the current context, I would argue that it actually fits into GAP.jl.)